### PR TITLE
MAINT: Improve shuffle performance in common case

### PR DESCRIPTION
**Pull request by bashtage**
Improve shuffle performance for C-contiguous arrays of any dimension

### CODE REVIEW

File: numpy/random/mtrand/mtrand.pyx [hunk 1]
```diff
@@ -4840,12 +4841,28 @@ cdef class RandomState:
...
-                    x[j] = x[i]
-                    x[i] = buf
+                    if fast_buffer_copy:
+                        memcpy(buf_ptr, x_ptr + j * buffer_size, buffer_size)
+                        memcpy(x_ptr + j * buffer_size,
+                               x_ptr + i * buffer_size,
+                               buffer_size)
+                        memcpy(x_ptr + i * buffer_size,
+                               buf_ptr,
+                               buffer_size)
```
Comment (eric-wieser) [inline numpy/random/mtrand/mtrand.pyx]: Not all dtypes are `memcpy`able, as @EricCousineau-TRI will tell you

Comment (bashtage) [reply numpy/random/mtrand/mtrand.pyx]: That was a question I had too.

Comment (bashtage) [reply numpy/random/mtrand/mtrand.pyx]: Is there a reasonable method to filter these cases?

Comment (EricCousineau-TRI) [reply numpy/random/mtrand/mtrand.pyx]: > Not all dtypes are memcpyable, as @EricCousineau-TRI will tell you

Yuppers, captured in #10898; also just filed #11076 regarding sensitivity to memory initialization as well.

> Is there a reasonable method to filter these cases?

I think the intent post-#10898 may be to have the `REFCHK` flag more generally mean non-trivially copyable, but I myself cannot suggest anything useful for the present case...
Should we consider drafting some WIP attributes to indicate this, solidify them in future versions?

**Comment by pv** [comment]
I guess you could check if the type code is one of the Numpy basic
types.

**Comment by bashtage** [comment]
One follow up question:

> In this case, AutoDiffXd (Eigen::AutoDiffScalar<Eigen::VectorXd>) has its value() stored immediately in the structure (which is trivially copyable), while its derivatives() are stored using Eigen::VectorXd, which is not trivially copyable since it has a pointer for its allocated data, and thus can't be memcpyd as the memory will be incorrectly aliased (and will later cause a double-free if the original and the copy have their destructors called or are assigned over).

The issue here seems to be that a copy can lead to untracked memory when is a pointer.  This is an in-place operation and not a real copy.  It seems like particular issue isn't applicable to inplace operations.  Does this make sense?

**Comment by pv** [comment]
A nasty case:
```
struct {
   char buf[8];
   int *p;
}
```
where `p` may on initialization be made to point to `buf` --- such
structures are not memcpy safe.

**Comment by bashtage** [comment]
I can see that. I don't think that the existing 1d shuffle will work here either since it is doing a memcpy. 

https://github.com/numpy/numpy/blob/master/numpy/random/mtrand/mtrand.pyx#L4862

**Comment by EricCousineau-TRI** [comment]
> [...] It seems like particular issue isn't applicable to inplace operations. Does this make sense?

Ah, yes, you are right! That's actually a constraint that we have placed on ourselves due to Eigen expecting types to be memcpy-movable too:
https://github.com/RobotLocomotion/drake/issues/5974
An example of testing it:
https://github.com/soonho-tri/drake/commit/41cc71e74c563d9ea764acfb39b87dd45825773b
We explicitly avoid the nasty case above, as this is what `std::string` was doing due to SSO.

So yes, if it is just moving memory, and not attempting to copy memory, then all should be good for our current use case. At some point in the future, it'd be nice to permit non-memcpy movable types, but it's not a strong requirement.

**Comment by bashtage** [comment]
It is now clear that this method will work in any case where the 1d shuffle would work since it uses the same inline memcpy swap.

**Comment by eric-wieser** [comment]
Can you give the benchmark results?

I'm not a huge fan of this optimization - it reduces generality, and if there's really a problem it feels like we should be optimizing `x[...] = y` instead.

**Comment by bashtage** [comment]
From branch:
```
In [1]: import numpy as np
In [2]: x=np.empty((50000,2))
In [3]: %timeit np.random.shuffle(x)
1000 loops, best of 3: 1.34 ms per loop
```

1.14.3
```
 %timeit np.random.shuffle(x)
10 loops, best of 3: 45.6 ms per loop
```

so 1.5 OOM faster.

The existing method is essentially pure Python which is why it is slow.

**Comment by bashtage** [comment]
As far as the case against, I can see not wanting to have lots of special cases.  Currently, the 1d array is special cased and everything else is pure Python.  If there was room for only 1 special case, I woudl argue that it should be for any C contiguous array for any shape rather than strided 1-d arrays.  IMO using shufflle on a strided array (esp a view of another array) seems a bit dumb, althoguh I'm sure someone somewhere has a legit use for this feature. 

Striding makes writing a fast general purpose array difficult.  It would be great if there was a low level function somewhere that could do a strided copy to an existing array without returning to Python.

**Comment by bashtage** [comment]
Closing since makes tradeoffs and not worth the bandwidth now.

**Pull request closed without merging**